homeassistant/components/xiaomi_miio/humidifier.py

```diff
@@ -0,0 +1,660 @@
+    MODEL_AIRHUMIDIFIER_CA4,
+    MODEL_AIRHUMIDIFIER_CB1,
+    MODELS_HUMIDIFIER,
+    MODELS_HUMIDIFIER_MIOT,
+    SERVICE_HUMIDIFIER_SET_BUZZER_OFF,
+    SERVICE_HUMIDIFIER_SET_BUZZER_ON,
+    SERVICE_HUMIDIFIER_SET_CHILD_LOCK_OFF,
+    SERVICE_HUMIDIFIER_SET_CHILD_LOCK_ON,
+    SERVICE_HUMIDIFIER_SET_DRY_OFF,
+    SERVICE_HUMIDIFIER_SET_DRY_ON,
+    SERVICE_HUMIDIFIER_SET_LED_BRIGHTNESS,
+    SERVICE_HUMIDIFIER_SET_LED_OFF,
+    SERVICE_HUMIDIFIER_SET_LED_ON,
+    SERVICE_HUMIDIFIER_SET_MOTOR_SPEED,
+    SERVICE_HUMIDIFIER_SET_TARGET_HUMIDITY,
+)
+from .device import XiaomiMiioEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Xiaomi Miio Device"
+DEFAULT_MIN_HUMIDITY = 30
+DEFAULT_MAX_HUMIDITY = 80
+DATA_KEY = "humidifier.xiaomi_miio"
+
+CONF_MODEL = "model"
+
+ATTR_MODEL = "model"
+
+MODE_AUTO = "Auto"
+MODE_HIGH = "High"
+MODE_LOW = "Low"
+MODE_MEDIUM = "Medium"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_TOKEN): vol.All(cv.string, vol.Length(min=32, max=32)),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_MODEL): vol.In(MODELS_HUMIDIFIER),
+    }
+)
+
+ATTR_TEMPERATURE = "temperature"
+ATTR_HUMIDITY = "humidity"
+ATTR_BUZZER = "buzzer"
+ATTR_CHILD_LOCK = "child_lock"
+ATTR_LED = "led"
+ATTR_LED_BRIGHTNESS = "led_brightness"
+ATTR_BRIGHTNESS = "brightness"
+ATTR_LEVEL = "level"
+ATTR_CURRENT_HUMIDITY = "current_humidity"
+ATTR_FEATURES = "features"
+ATTR_VOLUME = "volume"
+ATTR_USE_TIME = "use_time"
+ATTR_BUTTON_PRESSED = "button_pressed"
+
+# Air Humidifier
+ATTR_TARGET_HUMIDITY = "target_humidity"
+ATTR_TRANS_LEVEL = "trans_level"
+ATTR_HARDWARE_VERSION = "hardware_version"
+
+# Air Humidifier CA
+ATTR_MOTOR_SPEED = "motor_speed"
+ATTR_DEPTH = "depth"
+ATTR_DRY = "dry"
+
+# Air Humidifier CA4
+ATTR_ACTUAL_MOTOR_SPEED = "actual_speed"
+ATTR_FAHRENHEIT = "fahrenheit"
+ATTR_FAULT = "fault"
+ATTR_WATER_LEVEL = "water_level"
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON = {
+    ATTR_TEMPERATURE: "temperature",
+    ATTR_CURRENT_HUMIDITY: "humidity",
+    ATTR_BUZZER: "buzzer",
+    ATTR_CHILD_LOCK: "child_lock",
+    ATTR_LED_BRIGHTNESS: "led_brightness",
+    ATTR_USE_TIME: "use_time",
+}
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER = {
+    **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_TRANS_LEVEL: "trans_level",
+    ATTR_BUTTON_PRESSED: "button_pressed",
+    ATTR_HARDWARE_VERSION: "hardware_version",
+}
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA_AND_CB = {
+    **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_MOTOR_SPEED: "motor_speed",
+    ATTR_DEPTH: "depth",
+    ATTR_DRY: "dry",
+    ATTR_HARDWARE_VERSION: "hardware_version",
+}
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA4 = {
+    **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_ACTUAL_MOTOR_SPEED: "actual_speed",
+    ATTR_BUTTON_PRESSED: "button_pressed",
+    ATTR_DRY: "dry",
+    ATTR_FAHRENHEIT: "fahrenheit",
```

Comment:
Temperature and water level and similar measurements that we don't control should be separate sensor entities and not state attributes of the humidifier.

Reply:
Could we address the attribute issues in a later refactoring?

Reply:
Could we keep `temperature`, `water_level` and `depth` (same as water level for some models) attributes to not create breaking change and as @syssi proposed address moving those values to sensors in the next PR?

Reply:
It's a new platforms so it's already a breaking change. It's a good opportunity to clean up.

I propose we don't release this PR until we've moved these measurements to a sensor platform in a separate PR.

Reply:
So I convert this PR to the draft and I'll try to add `sensor` platform for humidifiers.

Reply:
Please coordinate with @starkillerOG and https://github.com/home-assistant/core/pull/46964.

Reply:
See this post: https://github.com/home-assistant/core/pull/46964#discussion_r587736441
Splitting up in seperate enties is beeing worked on but will require significant more work.
See https://github.com/rytilahti/python-miio/issues/929#issuecomment-787442355 if you want to help with this